README.md
# buffer-static-upload

[![Build Status](https://travis-ci.com/bufferapp/buffer-static-upload.svg?branch=master)](https://travis-ci.com/bufferapp/buffer-static-upload)

A straightforward static asset uploader which versions files by their contents
and generates a json file for your application to read the uploaded file
locations from.

## Versioned files and Images

`.js` and `.css` files are versioned during upload using a hash of the file's
contents as to allow for cache-busting. You can skip the automatic versioning
by using the `-skip-versioning` flag.

Images and other files are *not* versioned to allow for maximum caching and due
to their contents not changing very often like `.css` and `.js` files do.

## Install

A pre-compiled binary is available for download for both Linux and macOS.
Replace the version (ex. `0.2.0`) in the URL below for the version you require:

```
curl -L https://github.com/bufferapp/buffer-static-upload/releases/download/0.2.0/buffer-static-upload-`uname -s` > /usr/local/bin/buffer-static-upload
chmod +x /usr/local/bin/buffer-static-upload
```

## Usage

Ensure your AWS credentials environment variables are set (`AWS_ACCESS_KEY_ID`,
`AWS_SECRET_ACCESS_KEY`). The cli has the following argument options:

```
$ buffer-static-upload -h
Usage of ./buffer-static-upload:
  -bucket string
    	the s3 bucket to upload to (default "static.buffer.com")
  -dir string
    	required, the directory to upload files to in the bucket
  -dry-run
    	print the output only, skip file uploads and manifest creation
  -files string
    	the path to the files you'd like to upload, ex. "public/**/.*js,public/style.css"
  -format string
    	format of the output [json,csv] (default "json")
  -o string
    	the filename for the versions manifest (default "staticAssets.json")
  -skip-versioning
    	skip versioning uploaded files
  -v	print the current buffer-static-upload version
```

For example, you can use glob patterns to match multiple sets of files:

```
buffer-static-upload -files "public/js/**/*.js,public/css/*.css,public/img/*.*" -bucket my-bucket
```

This will generate a `staticAssets.json` file in this directory like this:

```json
{
  "public/css/style.css": "https://my-bucket.s3.amazonaws.com/public/css/style.11985b07e3121564a73d4d6821bfcfe7.css",
  "public/js/x/another.js": "https://my-bucket.s3.amazonaws.com/public/js/x/another.bfa2d0f60841707efe7be0a94c4caacf.js",
  "public/js/script.js": "https://my-bucket.s3.amazonaws.com/public/js/script.d55002b60fcfff0b3d355184d23af6f7.js",
  "public/img/home.jpg": "https://my-bucket.s3.amazonaws.com/public/img/home.jpg",
}
```

*Note* - The default bucket is used by multiple teams, so if you use that you
must specify a directory to use for your project as not to create unnecessary
collisions.

### Development

To work on this project you'll need [Golang](https://golang.org/dl/) and
the [Glide package manager](https://glide.sh/) installed. You should have
your `GOPATH` environment variable set and this repo should be cloned within
your `$GOPATH/src/github.com/bufferapp` directory.

To install the dependencies, run:

```
glide install
```

To test the script run:

```
go run main.go <your cli arguments here>
```

When distributing a new release version, run this script to generate the
binaries for Linux and Mac:

```
./build.sh
```

### License

MIT


CHANGELOG.md
0.3.0

- Add `skip-versioning` flag to disable automatic hash-based versioning for js & css files
- Fix bug where the upload would fail if a glob pattern matched a directory

0.2.0

- Add `format` flag for json or csv manifest output format
- Add `dry-run` flag for testing configurations

0.1.0

- Upload and version files to s3 bucket
- Generate `json` manifest file with all uploaded file locations


main_test.go
package main

import (
	"fmt"
	"os"
	"testing"

	"github.com/spf13/afero"
)

func TestGetFileURL(t *testing.T) {
	tests := []struct {
		bucket         string
		bucketFilename string
		expected       string
	}{
		{"some-bucket", "js/app.1234.js", "https://s3.amazonaws.com/some-bucket/js/app.1234.js"},
		{"static.buffer.com", "dir/js/app.1234.js", "https://static.buffer.com/dir/js/app.1234.js"},
	}

	for _, test := range tests {
		actual := GetFileURL(test.bucket, test.bucketFilename)
		if actual != test.expected {
			t.Errorf("File URL was incorrect, got: %s, expected %s", actual, test.expected)
		}
	}
}

func TestShouldVersionFile(t *testing.T) {
	tests := []struct {
		filename string
		expected bool
	}{
		{"bundle.js", true},
		{"assets.css", true},
		{"another.file", false},
	}

	for _, test := range tests {
		actual := ShouldVersionFile(test.filename)
		if actual != test.expected {
			t.Errorf("ShouldVersionFile result was incorrect, got %t, expected %t for filename %s", actual, test.expected, test.filename)
		}
	}
}

func TestGetUploadFilename(t *testing.T) {
	var AppFs = afero.NewMemMapFs()
	filename := "bundle.js"
	file, _ := AppFs.OpenFile(filename, os.O_CREATE, 0600)
	file.WriteString("some JS content")
	tests := []struct {
		file           afero.File
		filename       string
		skipVersioning bool
		expected       string
	}{
		{file, filename, false, "bundle.d41d8cd98f00b204e9800998ecf8427e.js"},
		{file, filename, true, "bundle.js"},
	}

	for _, test := range tests {
		fmt.Print(file.Name())
		actual, _ := GetUploadFilename(test.file, test.filename, test.skipVersioning)
		if actual != test.expected {
			t.Errorf("GetUploadFilename result was incorrect, got %s, expected %s", actual, test.expected)
		}
	}
}


tests/utils_test.go
package utils

import (
	"os"
	"testing"

	"github.com/bufferapp/buffer-static-upload/utils"
)

func TestGetFileMd5(t *testing.T) {
	file, _ := os.Open("hash-test.txt")
	hash, _ := utils.GetFileMd5(file)
	expected := "bde181f4a40d6e56662537d4e643a487"
	if hash != expected {
		t.Error("Expected", expected, "got", hash)
	}
}


utils/utils.go
package utils

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"io"
	"mime"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/endpoints"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3/s3manager"
)

// GetFileMd5 returns a checksum for a given file
func GetFileMd5(file *os.File) (string, error) {
	var fileHash string
	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		return fileHash, err
	}
	hashInBytes := hash.Sum(nil)[:16]
	fileHash = hex.EncodeToString(hashInBytes)
	return fileHash, nil
}

// GetVersionedFilename returns a new filename with the version before the extension
func GetVersionedFilename(filename string, version string) string {
	ext := filepath.Ext(filename)
	versionedExt := "." + version + ext
	versionedFilename := strings.Replace(filename, ext, versionedExt, 1)
	return versionedFilename
}

// GetFileMimeType returns the mime type of a file using it's extension
func GetFileMimeType(filename string) string {
	ext := filepath.Ext(filename)
	return mime.TypeByExtension(ext)
}

// GetFilesFromGlobsList returns a list of files that match a list of
// comma-deliniated file path globs
func GetFilesFromGlobsList(globList string) ([]string, error) {
	var files []string
	globs := strings.Split(globList, ",")

	for _, glob := range globs {
		fileList, err := filepath.Glob(glob)
		if err != nil {
			return files, err
		}
		files = append(files, fileList...)
	}
	return files, nil
}

// GetS3Uploader returns a configured Uploader
func GetS3Uploader() (*s3manager.Uploader, error) {
	var uploader *s3manager.Uploader

	awsAccessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
	awsSecretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")

	creds := credentials.NewStaticCredentials(awsAccessKeyID, awsSecretAccessKey, "")

	sess := session.Must(session.NewSession(&aws.Config{
		Credentials: creds,
		Region:      aws.String(endpoints.UsEast1RegionID),
	}))

	_, err := creds.Get()
	if err != nil {
		fmt.Printf("Bad credentials: %s", err)
		return uploader, err
	}

	uploader = s3manager.NewUploader(sess)
	return uploader, nil
}

// VersionAndUploadFiles will verion files and upload them to s3 and return
// a map of filenames and their version hashes
func VersionAndUploadFiles(bucket string, filenames []string) (map[string]string, error) {
	fileVersions := map[string]string{}

	uploader, err := GetS3Uploader()
	if err != nil {
		return fileVersions, err
	}

	for _, filename := range filenames {
		file, err := os.Open(filename)
		if err != nil {
			return fileVersions, err
		}
		defer file.Close()

		checksum, err := GetFileMd5(file)
		if err != nil {
			return fileVersions, err
		}

		fileVersions[filename] = checksum
		versionedFilename := GetVersionedFilename(filename, checksum)
		mimeType := GetFileMimeType(filename)

		result, err := uploader.Upload(&s3manager.UploadInput{
			Bucket:       aws.String(bucket),
			Key:          aws.String(versionedFilename),
			ContentType:  aws.String(mimeType),
			CacheControl: aws.String("public, max-age=31520626"),
			Expires:      aws.Time(time.Now().AddDate(10, 0, 0)),
			Body:         file,
		})
		if err != nil {
			return fileVersions, err
		}

		fmt.Printf("Uploaded %s\n", result.Location)
	}

	return fileVersions, nil
}


main.go
package main

import (
	"bytes"
	"crypto/md5"
	"encoding/csv"
	"encoding/hex"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"mime"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/endpoints"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/s3/s3manager"
)

// VERSION is the current version
const VERSION = "0.2.1"

var defaultS3Bucket = "static.buffer.com"
var uploader *s3manager.Uploader
var svc *s3.S3

func fatal(format string, a ...interface{}) {
	s := "Error: " + format + "\n"
	if a != nil {
		fmt.Printf(s, a)
	} else {
		fmt.Print(s)
	}
	os.Exit(1)
}

// GetFileMd5 returns a checksum for a given file
func GetFileMd5(file io.Reader) (string, error) {
	var fileHash string
	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		return fileHash, err
	}
	hashInBytes := hash.Sum(nil)[:16]
	fileHash = hex.EncodeToString(hashInBytes)
	return fileHash, nil
}

// GetVersionedFilename returns a new filename with the version before the extension
func GetVersionedFilename(filename string, version string) string {
	ext := filepath.Ext(filename)
	versionedExt := "." + version + ext
	versionedFilename := strings.Replace(filename, ext, versionedExt, 1)
	return versionedFilename
}

// GetFileMimeType returns the mime type of a file using it's extension
func GetFileMimeType(filename string) string {
	ext := filepath.Ext(filename)
	return mime.TypeByExtension(ext)
}

// GetFilesFromGlobsList returns a list of files that match a list of
// comma-deliniated file path globs
func GetFilesFromGlobsList(globList string) ([]string, error) {
	var files []string
	globs := strings.Split(globList, ",")

	for _, glob := range globs {
		fileList, err := filepath.Glob(glob)
		if err != nil {
			return files, err
		}
		files = append(files, fileList...)
	}
	return files, nil
}

// SetupS3Uploader configures and assigns the global "uploader" and "svc" variables
func SetupS3Uploader() {
	awsAccessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
	awsSecretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")

	creds := credentials.NewStaticCredentials(awsAccessKeyID, awsSecretAccessKey, "")

	sess := session.Must(session.NewSession(&aws.Config{
		Credentials: creds,
		Region:      aws.String(endpoints.UsEast1RegionID),
	}))
	_, err := creds.Get()
	if err != nil {
		fatal("failed to load AWS credentials %s", err)
	}

	uploader = s3manager.NewUploader(sess)
	svc = s3.New(sess)
}

// HasPreviousUpload performs a HEAD request to check if a file has been uploaded already
func HasPreviousUpload(svc *s3.S3, bucket string, filename string) bool {
	req, _ := svc.HeadObjectRequest(&s3.HeadObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(filename),
	})
	err := req.Send()
	if err == nil {
		return true
	}
	return false
}

// GetFileURL returns the final url of the file
func GetFileURL(bucket string, bucketFilename string) string {
	// the static.buffer.com bucket has a domain alias
	if bucket == defaultS3Bucket {
		return "https://" + path.Join(bucket, bucketFilename)
	}
	return "https://s3.amazonaws.com" + path.Join("/", bucket, "/", bucketFilename)
}

// UploadFile uploads a given file to the s3 bucket
func UploadFile(file *os.File, filename string, bucket string) (err error) {
	mimeType := GetFileMimeType(filename)
	_, err = uploader.Upload(&s3manager.UploadInput{
		Bucket:       aws.String(bucket),
		Key:          aws.String(filename),
		ContentType:  aws.String(mimeType),
		CacheControl: aws.String("public, max-age=31520626"),
		Expires:      aws.Time(time.Now().AddDate(10, 0, 0)),
		Body:         file,
	})
	if err != nil {
		return err
	}
	return nil
}

func ShouldVersionFile(filename string) bool {
	ext := filepath.Ext(filename)
	return ext == ".js" || ext == ".css"
}

func GetUploadFilename(file io.Reader, filename string, skipVersioning bool) (string, error) {
	uploadFilename := filename
	if !skipVersioning && ShouldVersionFile(filename) {
		checksum, errMd5 := GetFileMd5(file)
		if errMd5 != nil {
			return "", errMd5
		}
		uploadFilename = GetVersionedFilename(filename, checksum)
	}
	return uploadFilename, nil
}

func IsDirectory(path string) (bool, error) {
	info, err := os.Stat(path)
	if err != nil {
		return false, err
	}
	return info.IsDir(), err
}

// VersionAndUploadFiles will verion files and upload them to s3 and return
// a map of filenames and their version hashes
func VersionAndUploadFiles(
	bucket string,
	directory string,
	filenames []string,
	dryRun bool,
	skipVersioning bool,
) (map[string]string, error) {
	fileVersions := map[string]string{}

	fmt.Printf("Uploading to %s/%s\n", bucket, directory)

	for _, filename := range filenames {
		isDir, err := IsDirectory(filename)
		if isDir || err != nil {
			continue
		}
		file, err := os.Open(filename)
		if err != nil {
			return fileVersions, err
		}
		defer file.Close()

		uploadFilename, err := GetUploadFilename(file, filename, skipVersioning)
		if err != nil {
			return fileVersions, err
		}
		bucketFilename := path.Join(directory, uploadFilename)
		fileURL := GetFileURL(bucket, bucketFilename)

		shouldUpload := !HasPreviousUpload(svc, bucket, bucketFilename)
		if shouldUpload && !dryRun {
			err := UploadFile(file, bucketFilename, bucket)
			if err != nil {
				return fileVersions, err
			}
		}

		if shouldUpload {
			fmt.Printf("%-10s %s\n", "Uploaded", filename)
		} else {
			fmt.Printf("%-10s %s\n", "Skipped", filename)
		}

		fileVersions[filename] = fileURL
	}

	return fileVersions, nil
}

// FormatManifest returns the file version manifest in json or csv format
func FormatManifest(fileVersions map[string]string, format string) ([]byte, error) {
	if format == "json" {
		return json.MarshalIndent(fileVersions, "", "  ")
	}
	if format == "csv" {
		b := &bytes.Buffer{}
		wr := csv.NewWriter(b)
		for filename, uri := range fileVersions {
			row := []string{filename, uri}
			wr.Write(row)
		}
		wr.Flush()
		return b.Bytes(), nil
	}
	return nil, nil
}

func main() {
	s3Bucket := flag.String("bucket", defaultS3Bucket, "the s3 bucket to upload to")
	directory := flag.String("dir", "", "required, the directory to upload files to in the bucket")
	filesArg := flag.String("files", "", "the path to the files you'd like to upload, ex. \"public/**/.*js,public/style.css\"")
	outputFilename := flag.String("o", "staticAssets.json", "the filename for the versions manifest")
	format := flag.String("format", "json", "format of the output [json,csv]")
	dryRun := flag.Bool("dry-run", false, "print the output only, skip file uploads and manifest creation")
	printVersion := flag.Bool("v", false, "print the current buffer-static-upload version")
	skipVersioning := flag.Bool("skip-versioning", false, "skip versioning uploaded files")
	flag.Parse()

	if *printVersion {
		fmt.Printf("%s\n", VERSION)
		os.Exit(0)
	}

	if *directory == "" && *s3Bucket == defaultS3Bucket {
		fatal("To use the default bucket you need to specify an upload directory (-dir)")
	}

	start := time.Now()
	files, err := GetFilesFromGlobsList(*filesArg)
	if err != nil {
		fatal("failed to get files %s", err)
	}
	append := " and version"
	if *skipVersioning == true {
		append = ""
	}
	fmt.Printf("Found %d files to upload%s:\n", len(files), append)

	SetupS3Uploader()
	fileVersions, err := VersionAndUploadFiles(*s3Bucket, *directory, files, *dryRun, *skipVersioning)
	if err != nil {
		fatal("failed to upload files %s", err)
	}

	output, err := FormatManifest(fileVersions, *format)
	if err != nil {
		fatal("failed to format versions manifest file %s", err)
	}

	if !*dryRun {
		err = ioutil.WriteFile(*outputFilename, output, 0644)
		if err != nil {
			fatal("failed to write versions mainifest file %s", err)
		}
	}

	elapsed := time.Since(start)
	if *dryRun {
		fmt.Printf("\nCompleted dry run in %s\n", elapsed)
	} else {

		fmt.Printf(
			"\nSuccessfully uploaded static assets and generated %s in %s\n",
			*outputFilename,
			elapsed,
		)
	}
}
